src/main/java/org/skriptlang/skript/lang/script/Annotation.java

```diff
@@ -0,0 +1,183 @@
+ * Some may have a global effect on any syntax (e.g. suppressing a warning, changing parsing order)
+ * whereas others might be relevant ONLY to one syntax.
+ * <p>
+ * This interface represents a post-parsing annotation.
+ */
+public interface Annotation extends CharSequence {
+
+	static @NotNull Annotation create(String text) {
+		return new SimpleAnnotation(text);
+	}
+
+	static boolean isAnnotationPresent(String text) {
+		return ParserInstance.get().hasAnnotation(text);
+	}
+
+	static boolean isAnnotationPresent(SkriptPattern pattern) {
+		return ParserInstance.get().hasAnnotationMatching(pattern);
+	}
+
+	/**
+	 * @return The annotation's content after the initial '@'
+	 */
+	@NotNull
+	String value();
+
+	/**
+	 * Whether the content of this annotation equals the given value.
+	 */
+	default boolean valueEquals(@Nullable String value) {
+		return this.value().equals(value);
+	}
+
+	class Match implements Annotation {
+
+		private final Annotation annotation;
+		private final MatchResult result;
+
+		public Match(Annotation annotation, MatchResult result) {
+			this.annotation = annotation;
+			this.result = result;
+		}
+
+		public MatchResult result() {
+			return result;
+		}
+
+		public Annotation annotation() {
+			return annotation;
+		}
+
+		@Override
+		public @NotNull String value() {
+			return annotation.value();
+		}
+
+		@Override
+		public int length() {
+			return annotation.length();
+		}
+
+		@Override
+		public char charAt(int index) {
+			return annotation.charAt(index);
+		}
+
+		@NotNull
+		@Override
+		public CharSequence subSequence(int start, int end) {
+			return annotation.subSequence(start, end);
+		}
+
+		@Override
+		public String toString() {
+			return annotation.toString();
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (obj instanceof Annotation)
+				return Objects.equals(this.value(), ((Annotation) obj).value())
+					|| annotation.equals(obj);
+			else
+				return false;
+		}
+
+	}
+
+}
+
+final class SimpleAnnotation implements Annotation, CharSequence {
+
+	private final String value;
+
+	SimpleAnnotation(@NotNull String value) {
+		this.value = value;
+	}
+
+	@Override
+	public @NotNull String value() {
+		return value;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (obj instanceof Annotation)
+			return Objects.equals(value, ((Annotation) obj).value());
+		else
+			return false;
+	}
+
+	@Override
+	public @NotNull String toString() {
+		return '@' + value;
+	}
+
+	@Override
+	public int hashCode() {
+		return value.hashCode();
+	}
+
+	@Override
+	public int length() {
+		return 1 + value.length();
+	}
+
+	@Override
+	public char charAt(int index) {
+		if (index == 0)
+			return '@';
+		return value.charAt(index - 1);
+	}
+
+	@Override
+	public String subSequence(int start, int end) {
+		if (start == 0)
+			return '@' + value.substring(0, end + 1);
+		return value.substring(start + 1, end + 1);
```

Comment:
Is this right? If I do `annotation.subSequence(1, <end>)`, it should give everything after the `@`. This appears to skip forward instead of backward (at least for `start`)

Reply:
I don't know, I don't really remember how this works 😬

Reply:
Yeah, I think there needs to be subtraction here rather than addition for `start`. I also think you want to leave `end` as-is. If for some annotation `@hello`, if I do `annotation.subSequence(0, 2)`, that should give `@h`, but it would currently give `@he` (`end` should be exclusive)